README.md
# @macherjek/storyblok-schedule

[![npm version](https://img.shields.io/npm/v/@kshevitskiy/storyblok-schedule)](https://www.npmjs.com/package/@kshevitskiy/storyblok-schedule)

**Component to display weekly schedule**

This compenent could be used with [storyblok opening hours](https://www.notion.so/OpeningHours-Component-41cadbfd790749c29dbe535af1097de0) plugin or with custom data passed as `days` property.

See [docs](https://github.com/) for more information.

## Setup CI environent variables

1. Login to your npm account
   ```sh
   npm login
   ```
2. [Create](https://docs.npmjs.com/creating-and-viewing-access-tokens) and copy npm access token
   ```sh
   npm token create
   ```
3. Go to GitHub repo: Settings > Secrets > New repository secret:
   - name: `NPM_AUTH_TOKEN`
   - value: token

**NOTICE. Package will be published in npm and [GitHub package registry](https://github.com/features/packages)**

## License

[MIT Licensed](license) by [Macherjek](https://www.macherjek.at/)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.25](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.24...v1.0.25) (2021-12-14)

### [1.0.24](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.23...v1.0.24) (2021-11-02)

### [1.0.23](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.22...v1.0.23) (2021-11-02)

### [1.0.22](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.21...v1.0.22) (2021-08-31)

### [1.0.21](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.20...v1.0.21) (2021-08-31)

### [1.0.20](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.19...v1.0.20) (2021-08-31)

### [1.0.19](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.18...v1.0.19) (2021-07-16)

### [1.0.18](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.17...v1.0.18) (2021-05-03)

### [1.0.17](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.16...v1.0.17) (2021-05-03)

### [1.0.16](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.15...v1.0.16) (2021-04-19)

### [1.0.15](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.14...v1.0.15) (2021-04-19)

### [1.0.14](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.13...v1.0.14) (2021-04-19)

### [1.0.13](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.12...v1.0.13) (2021-04-19)

### [1.0.12](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.11...v1.0.12) (2021-04-19)

### [1.0.11](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.10...v1.0.11) (2021-04-19)

### [1.0.10](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.9...v1.0.10) (2021-04-18)

### [1.0.9](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.8...v1.0.9) (2021-04-18)

### [1.0.8](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.7...v1.0.8) (2021-04-18)

### [1.0.7](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.6...v1.0.7) (2021-04-18)

### [1.0.6](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.5...v1.0.6) (2021-04-18)

### [1.0.5](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.4...v1.0.5) (2021-04-18)

### [1.0.4](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.3...v1.0.4) (2021-04-18)

### [1.0.3](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.2...v1.0.3) (2021-04-18)

### [1.0.2](https://github.com/kshevitskiy/storyblok-schedule/compare/v1.0.1...v1.0.2) (2021-04-18)

### 1.0.1 (2021-04-18)


src/storyblok-schedule.test.js
import { shallowMount } from '@vue/test-utils'
import StoryblokSchedule from './storyblok-schedule.vue'

test('Renders', () => {
  const wrapper = shallowMount(StoryblokSchedule)
  expect(wrapper.exists()).toBe(true)
})

// test('Shows placeholders', () => {
//   const dayHtmlPlaceholder = '<div class="day"></div>'
//   const timeHtmlPlaceholder = '<div class="time"></div>'

//   const wrapper = shallowMount(StoryblokSchedule, {
//     slots: {
//       day: dayHtmlPlaceholder,
//       time: timeHtmlPlaceholder
//     }
//   })

//   expect(wrapper.html()).toContain(dayHtmlPlaceholder)
//   expect(wrapper.html()).toContain(timeHtmlPlaceholder)
// })


src/index.js
// Import vue component
import component from './storyblok-schedule.vue'

// install function executed by Vue.use()
function install(Vue) {
  if (install.installed) return
  install.installed = true
  Vue.component('StoryblokSchedule', component)
}

// Create module definition for Vue.use()
const plugin = {
  install
}

// To auto-install when vue is found
let GlobalVue = null
if (typeof window !== 'undefined') {
  GlobalVue = window.Vue
} else if (typeof global !== 'undefined') {
  GlobalVue = global.Vue
}
if (GlobalVue) {
  GlobalVue.use(plugin)
}

// Inject install function into component - allows component
// to be registered via Vue.use() as well as Vue.component()
component.install = install

// Export component by default
export default component

// It's possible to expose named exports when writing components that can
// also be used as directives, etc. - eg. import { RollupDemoDirective } from 'rollup-demo';
// export const RollupDemoDirective = component;


src/helpers.js
const TODAY = new Date()
const TIME_REGEX = /^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$/

/**
 * Returns list of localized days
 * locale: name of locale, eg. en-GB, default de-DE
 * weekday: formart of weekday short/long default short
 */
export const getWeekDaysForLocale = (locale = 'de-DE', weekday = 'short') => {
  const baseDate = new Date('1973-01-01')
  const weekDays = []

  for (let i = 0; i < 7; i++) {
    weekDays.push(baseDate.toLocaleDateString(locale, { weekday }))
    baseDate.setDate(baseDate.getDate() + 1)
  }

  return weekDays
}

/**
 * Current week-day number
 * Mon: 0, Tue: 1, Wed: 2, Thu: 3, Fri: 4, Sat: 5, Sun: 6
 */
export const today = (TODAY.getDay() + 6) % 7

/**
 * Group Array of JavaScript Objects by Key or Property Value
 */
export const groupBy = (key) => (array) =>
  array.reduce((objectsByKeyValue, obj) => {
    const value = obj[key]
    objectsByKeyValue[value] = (objectsByKeyValue[value] || []).concat(obj)
    return objectsByKeyValue
  }, {})

/**
 * Validate days sequence
 */
export const sequentialDays = (input) => {
  const rule = /(?:(?=01|12|23|34|45|56)\d)+\d/g
  const match = input.match(rule)
  return match !== null && input === match[0]
}

/**
 * Days property validation method
 * Array length and object properties validation
 */
export const daysValidator = (value) => {
  if (value.length !== 7) {
    console.warn('[storyblok-schedule]: Days property should contain 7 days')
    return false
  }

  if (value.some((item) => !item.name)) {
    console.warn(
      '[storyblok-schedule]: <Days>[day] `name` property value is required'
    )
    return false
  }

  if (value.some((item) => !item.times)) {
    console.warn(
      '[storyblok-schedule]: <Days>[day] `times` property is required'
    )
    return false
  }

  return true
}

/**
 * Convert 24hr time string into 12 hrs time format
 * @param {string} timeString
 * @param {string} locale
 */
export const toAmPmTime = (timeString, locale = 'en-US') => {
  const isTime = TIME_REGEX.test(timeString)

  if (isTime) {
    const dateTime = `1970-01-01T${timeString}Z`
    const date = new Date(dateTime)
    const settings = {
      timeZone: 'UTC',
      hour12: true,
      hour: 'numeric',
      minute: 'numeric'
    }
    return date.toLocaleTimeString(locale, settings)
  }

  return timeString
}


rollup.config.js
import vue from 'rollup-plugin-vue'
import buble from 'rollup-plugin-buble'
import commonjs from 'rollup-plugin-commonjs'
import replace from 'rollup-plugin-replace'
import { terser } from 'rollup-plugin-terser'
import resolve from 'rollup-plugin-node-resolve'
import css from 'rollup-plugin-css-only'
import minimist from 'minimist'
import pkg from './package.json'

const argv = minimist(process.argv.slice(2))

const baseConfig = {
  input: 'src/index.js',
  plugins: {
    preVue: [
      replace({
        'process.env.NODE_ENV': JSON.stringify('production')
      }),
      commonjs()
    ],
    vue: {
      css: true,
      template: {
        isProduction: true
      }
    },
    postVue: [
      buble({
        transforms: {
          dangerousForOf: true
        }
      })
    ]
  }
}

// Customize configs for individual targets
const buildFormats = []
if (!argv.format || argv.format === 'es') {
  const esConfig = {
    ...baseConfig,
    output: {
      file: pkg.module,
      format: 'esm',
      exports: 'named',
      sourcemap: true
    },
    plugins: [
      ...baseConfig.plugins.preVue,
      css({
        output: pkg.style
      }),
      vue({
        ...baseConfig.plugins.vue,
        css: false
      }),
      ...baseConfig.plugins.postVue,
      terser({
        output: {
          ecma: 6
        }
      }),
      resolve()
    ]
  }
  buildFormats.push(esConfig)
}

if (!argv.format || argv.format === 'cjs') {
  const umdConfig = {
    ...baseConfig,
    output: {
      compact: true,
      file: pkg.main,
      format: 'cjs',
      name: 'VueLazyLoad',
      exports: 'named',
      sourcemap: true
    },
    plugins: [
      ...baseConfig.plugins.preVue,
      css({
        output: pkg.style
      }),
      vue({
        ...baseConfig.plugins.vue,
        template: {
          ...baseConfig.plugins.vue.template,
          optimizeSSR: true
        },
        css: false
      }),
      ...baseConfig.plugins.postVue,
      resolve()
    ]
  }
  buildFormats.push(umdConfig)
}

if (!argv.format || argv.format === 'iife') {
  const unpkgConfig = {
    ...baseConfig,
    output: {
      compact: true,
      file: pkg.unpkg,
      format: 'iife',
      name: 'VueLazyLoad',
      exports: 'named',
      sourcemap: true
    },
    plugins: [
      ...baseConfig.plugins.preVue,
      vue(baseConfig.plugins.vue),
      ...baseConfig.plugins.postVue,
      terser({
        output: {
          ecma: 5
        }
      }),
      resolve()
    ]
  }
  buildFormats.push(unpkgConfig)
}

// Export config
export default buildFormats


src/seeder.js
export default {
  base: [
    {
      name: 'Monday',
      times: [
        {
          end: '12:00',
          start: '09:00'
        }
      ]
    },
    {
      name: 'Tuesday',
      times: [
        {
          end: '12:00',
          start: '09:00'
        }
      ]
    },
    {
      name: 'Wednesday',
      times: [
        {
          end: '18:00',
          start: '09:00'
        }
      ]
    },
    {
      name: 'Thursday',
      times: []
    },
    {
      name: 'Friday',
      times: [
        {
          end: '12:00',
          start: '09:00'
        }
      ]
    },
    {
      name: 'Saturday',
      times: [
        {
          end: '12:00',
          start: '09:00'
        }
      ]
    },
    {
      name: 'Sunday',
      times: [
        {
          end: '12:00',
          start: '09:00'
        }
      ]
    }
  ],
  days: [
    {
      name: 'Monday',
      text: '',
      times: [
        {
          end: '16:00',
          start: '08:00'
        }
      ],
      source: 'times'
    },
    {
      name: 'Tuesday',
      text: 'Ruhetag',
      times: [
        {
          end: '17:00',
          start: '08:00'
        }
      ],
      source: 'text'
    },
    {
      name: 'Wednesday',
      text: '',
      times: [
        {
          end: '17:00',
          start: '07:00'
        }
      ],
      source: 'times'
    },
    {
      name: 'Thursday',
      text: '',
      times: [
        {
          end: '17:00',
          start: '07:00'
        }
      ],
      source: 'times'
    },
    {
      name: 'Friday',
      text: '',
      times: [
        {
          end: '20:00',
          start: '09:00'
        }
      ],
      source: 'times'
    },
    {
      name: 'Saturday',
      text: '',
      times: [
        {
          end: '14:00',
          start: '10:00'
        }
      ],
      source: 'times'
    },
    {
      name: 'Sunday',
      text: 'Geschlossen',
      times: [
        {
          end: '17:00',
          start: '08:00'
        }
      ],
      source: 'text'
    }
  ]
}
